internal/controller/volumereplicationgroup_controller.go

```diff
@@ -803,44 +803,234 @@ func (v *VRGInstance) separatePVCsUsingVRGStatus(pvcList *corev1.PersistentVolum
+	peerClass *ramendrv1alpha1.PeerClass,
+	storageClass *storagev1.StorageClass,
+) *volrep.VolumeReplicationClass {
+	for _, replicationClass := range v.replClassList.Items {
+		rIDFromReplicationClass := replicationClass.GetLabels()[VolumeReplicationIDLabel]
+		sIDfromReplicationClass := replicationClass.GetLabels()[StorageIDLabel]
+
+		matched := sIDfromReplicationClass == storageClass.GetLabels()[StorageIDLabel] &&
+			rIDFromReplicationClass == peerClass.ReplicationID &&
+			replicationClass.Spec.Provisioner == storageClass.Provisioner
+
+		if matched {
+			return &replicationClass
+		}
+
+		continue
+	}
+
+	return nil
+}
+
+func (v *VRGInstance) findVolSnapClass(storageClass *storagev1.StorageClass,
+) (*snapv1.VolumeSnapshotClass, error) {
+	SnaphotClasses, err := v.volSyncHandler.GetVolumeSnapshotClasses()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, snapshotClass := range SnaphotClasses {
+		sIDFromSnapClass := snapshotClass.GetLabels()[StorageIDLabel]
+
+		matched := sIDFromSnapClass == storageClass.GetLabels()[StorageIDLabel] &&
+			snapshotClass.Driver == storageClass.Provisioner
+
+		if matched {
+			return &snapshotClass, nil
+		}
+
+		continue
+	}
+
+	return nil, nil
+}
+
+func (v *VRGInstance) validateAndGetStorageClass(scName *string, pvc *corev1.PersistentVolumeClaim,
+) (*storagev1.StorageClass, error) {
+	if scName == nil || *scName == "" {
+		return nil, fmt.Errorf("missing storage class name for PVC %s/%s", pvc.GetNamespace(), pvc.GetName())
+	}
+
+	storageClass := &storagev1.StorageClass{}
+	if err := v.reconciler.Get(v.ctx, types.NamespacedName{Name: *scName}, storageClass); err != nil {
+		return nil, fmt.Errorf("failed to get the storageclass with name %s (%w)", *scName, err)
+	}
+
+	return storageClass, nil
+}
+
+func (v *VRGInstance) findPeerClassMatchingSC(
+	storageClass *storagev1.StorageClass,
+	peerClasses []ramendrv1alpha1.PeerClass,
+	pvc *corev1.PersistentVolumeClaim,
+) (*ramendrv1alpha1.PeerClass, error) {
+	var peerClass *ramendrv1alpha1.PeerClass
+
+	if _, ok := storageClass.GetLabels()[StorageIDLabel]; !ok {
+		return nil, fmt.Errorf("label (%s) not found in storageClass for PVC %s", StorageIDLabel, pvc.Name)
+	}
+
+	for idx := range peerClasses {
+		if storageClass.GetName() == peerClasses[idx].StorageClassName {
+			peerClass = &peerClasses[idx]
+		}
+	}
+
+	if peerClass == nil {
+		msg := fmt.Sprintf("peerClass matching storageClass %s not found for PVC", storageClass.GetName())
+		v.updatePVCDataReadyCondition(pvc.Namespace, pvc.Name, VRGConditionReasonPeerClassNotFound, msg)
+
+		return nil, fmt.Errorf(msg)
+	}
+
+	if !slices.Contains(peerClass.StorageID, storageClass.GetLabels()[StorageIDLabel]) {
+		return nil, fmt.Errorf("storageID mismatch between peerClass (%s) and StorageClass (%s)",
+			peerClass.StorageID[0], storageClass.GetLabels()[StorageIDLabel])
+	}
+
+	/* Removed as a comment, as tests set the same sID across both SCs and hence in async we get len as 1 always!
```

Comment:
Should delete this comment.

Reply:
We should track this and fix it in tests as well (at which point this code becomes active).